### PR TITLE
Scheduler refactor

### PR DESCRIPTION
**Pull request by csarofeen**
Overhauling of schedulers again. Overall changes are overwhelmingly positive though there are some cases that did regress.

image

More work could be done tuning, but I think this is good for now, will follow up with another PR that finishes channels first BN persistent kernels.

### CODE REVIEW

**Comment by csarofeen** [comment]
Needs https://github.com/csarofeen/pytorch/pull/1190 before merging.

File: test/cpp/jit/test_gpu.cpp [hunk 1]
```diff
@@ -9157,29 +9157,27 @@ TEST(NVFuserTest, FusionSmemDynamicTiledGemm_CUDA) {
...
+  // [Mo, No, Koo, Koi, Ki, Mi, Ni]
 
   // Factor out the outer reduction IterDomain, then run the inter-cta
   // reduction, and intra-cta reduction
   auto tv6 = tv5->rFactor({2});
+  // [Mo, No, rKoo, rKoi, rKi, Mi, Ni]
+  // [Mo, No,       rKoi, rKi, Mi, Ni]
 
   // Scope computations
   tv6->computeAt(tv5, 2);
```
Comment (csarofeen) [inline test/cpp/jit/test_gpu.cpp:9178]: Changed compute at ordering so it doesn't modify order if it doesn't have to during replay.

File: torch/csrc/jit/codegen/cuda/compute_at.cpp [hunk 2]
```diff
@@ -348,6 +348,97 @@ void ComputeAt::runWith(
   ca.runPass();
 }
 
+
+namespace {
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/compute_at.cpp:351]: Change compute at so it doesn't reorder axes if it doesn't need to, by avoiding running replayPasC and replayCasP if compute at that was called is satisfied in the current domains of tensors.

File: torch/csrc/jit/codegen/cuda/fusion_segmenter.cpp [hunk 3]
```diff
@@ -758,21 +758,31 @@ void SegmentedFusion::finalize() {
...
-  //  the group supports. Could consider moving this into
-  //  segmentedGroup in a follow up.
-  for (auto group : affected_group_set) {
-    auto input_group_vec = getAllInputs(group);
-    std::unordered_set<Val*> input_group_set(
-        input_group_vec.begin(), input_group_vec.end());
 
-    auto expr_set = DependencyCheck::getAllExprsBetween(
-        input_group_set, getAllOutputs(group));
-    group->exprs_ = std::vector<Expr*>(expr_set.begin(), expr_set.end());
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/fusion_segmenter.cpp:775]: There was a segfault here that didn't show up until I changed the schedulers. Need to update expressions as we process the groups.

File: torch/csrc/jit/codegen/cuda/fusion_segmenter.cpp [hunk 4]
```diff
@@ -3065,18 +3075,18 @@ class ForceHalfAnnotation : public IterVisitor {
     ForceHalfAnnotation annotation;
     std::vector<Val*> fp16_outputs;
     auto& cast_to_type = annotation.cast_to_type_;
+    auto other_half_type =
+        cast_to_type == DataType::Half ? DataType::BFloat16 : DataType::Half;
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/fusion_segmenter.cpp:3079]: Minor bf16 cleanup.

File: torch/csrc/jit/codegen/cuda/ir_utils.cpp [hunk 5]
```diff
@@ -295,9 +295,13 @@ TensorView* rfactorHelper(
 
   WelfordResult rtvs = reduction_tv->rFactor(axes, w_avg, w_var, w_n);
 
-  // TODO: this can be more generic, using avg because
-  //      WelfordOp::out() returns the avg
-  return rtvs.avg;
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/ir_utils.cpp:300]: Make rfactor helper behave more consistently by passing back the rfactor tensor based on what was passed in.

File: torch/csrc/jit/codegen/cuda/runtime/welford.cu [hunk 6]
```diff
@@ -15,9 +15,10 @@ __inline__ __device__ void welfordCombine(
     return;
   }
   TN ab_N = a_N + b_N;
+  T b_N_div_ab_N = ((T)(nvfuser_index_t)(b_N)) / ((T)(nvfuser_index_t)(ab_N));
   T delta = b_avg - a_avg;
-  a_avg += delta * b_N / ab_N;
-  a_M2 += b_M2 + delta * delta * a_N * b_N / ab_N;
+  a_avg += delta * b_N_div_ab_N;
+  a_M2 += b_M2 + delta * delta * ((T)(nvfuser_index_t)(a_N)) * b_N_div_ab_N;
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/runtime/welford.cu:21]: Be more careful about dtype during welford reduce.

File: torch/csrc/jit/codegen/cuda/scheduler/normalization.cpp [hunk 7]
```diff
@@ -5,6 +5,7 @@
 #include <torch/csrc/jit/codegen/cuda/ir_all_nodes.h>
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/scheduler/normalization.cpp:5]: Full rework of persistent kernels, move multi reduction logic out of normalization.cpp. This is now only for persistent kernel heuristics.

File: torch/csrc/jit/codegen/cuda/scheduler/pointwise.cpp [hunk 8]
```diff
@@ -67,32 +67,20 @@ c10::optional<PointwiseParams> getPointwiseHeuristics(
   if (TensorDomain::noReductions(
           TensorDomain::noBroadcasts(largest_out->domain()->domain()))
           .size() == 0) {
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/scheduler/pointwise.cpp:69]: There were issues in broadcast detection for 2D pointwise scheduling, much of this is to fix that.

File: torch/csrc/jit/codegen/cuda/scheduler/reduction.cpp [hunk 9]
```diff
@@ -4,6 +4,7 @@
 #include <torch/csrc/jit/codegen/cuda/instrumentation.h>
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/scheduler/reduction.cpp:4]: Rework reduction scheduling to be more inline with changes in normalization scheduler. Support multi reduction scheduling here.

Support "3D" schedules, i.e. reduction-iteration-reduction

File: torch/csrc/jit/codegen/cuda/scheduler/reduction_heuristic.h [hunk 10]
```diff
@@ -16,29 +16,81 @@ namespace cuda {
 class ReductionParams {
  public:
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/scheduler/reduction_heuristic.h:17]: Expand the heuristic structure to include more options so scheduling can be more incremental of a process. A lot of this came from supporting 3D reduction schedules, as well as having multiple unrolling dimensions or an unrolling and vectorization dimension. Not all arguments are cached, for example parallelization bindings are set here for convenience of rediscovering them. Other arguments were built out to be consistent in the options like for the outer reduction dim of a 3D schedule.

File: torch/csrc/jit/codegen/cuda/scheduler/reduction_utils.cpp [hunk 11]
```diff
@@ -0,0 +1,648 @@
+#include <torch/csrc/jit/codegen/cuda/scheduler/reduction_utils.h>
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/scheduler/reduction_utils.cpp:1]: Move these functions out of general scheduling utils and into its own file.

File: torch/csrc/jit/codegen/cuda/scheduler/reduction_utils.cpp [hunk 12]
```diff
@@ -0,0 +1,648 @@
...
+#include <torch/csrc/jit/codegen/cuda/transform_replay.h>
+
+namespace torch {
+namespace jit {
+namespace fuser {
+namespace cuda {
+
+namespace reduction_scheduler_utils {
+
+TensorView* scheduleReductionTV(
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/scheduler/reduction_utils.cpp:16]: Make this scheduling of the reduction dim more incremental.

File: torch/csrc/jit/codegen/cuda/scheduler/reduction_utils.cpp [hunk 13]
```diff
@@ -0,0 +1,648 @@
...
+  // Outer and inner reduction axis is relative. Outer reduce axis is only valid
+  // in 3D scheduling. Otherwise inner_reduce_axis is the only reduction axis.
+  // Inner here though is only relative to the other axis. When
+  // rparams.fastest_dim == false, the reduction axis is logically outside the
+  // iteration axis.
+  const int iter_axis = 0;
+  const int outer_reduce_axis = rparams.schedule_3D ? 1 : 0;
+  const int inner_reduce_axis = rparams.schedule_3D ? 2 : has_iter_axis ? 1 : 0;
+
+  TORCH_INTERNAL_ASSERT(
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/scheduler/reduction_utils.cpp:29]: Collection of sanity checks before scheduling.

File: torch/csrc/jit/codegen/cuda/scheduler/reduction_utils.cpp [hunk 14]
```diff
@@ -0,0 +1,648 @@
...
+
+    if (rparams.fastest_dim && rparams.split_grid_dim_iter_dom) {
+      reduction_tv->split(iter_axis, scheduler_utils::x_grid_limit);
+      reduction_tv->axis(iter_axis + 1)->parallelize(rparams.grid_dim_iter_dom);
+    } else {
+      reduction_tv->axis(iter_axis)->parallelize(rparams.grid_dim_iter_dom);
+    }
+  }
+
+  return sortAndRFactor(reduction_tv);
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/scheduler/reduction_utils.cpp:221]: This now automatically handles reorder + rfactor.

File: torch/csrc/jit/codegen/cuda/scheduler/reduction_utils.cpp [hunk 15]
```diff
@@ -0,0 +1,648 @@
...
+}
+
+void multiReductionInliner(
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/scheduler/reduction_utils.cpp:224]: Inliner that's coordinated with the ordering and structure of the tensors made by the above function.

File: torch/csrc/jit/codegen/cuda/scheduler/reduction_utils.cpp [hunk 16]
```diff
@@ -0,0 +1,648 @@
...
+
+      scheduler_utils::computeAtInputs(red_tv, pos, ComputeAtMode::MostInlined);
+      scheduler_utils::computeWithOutputs(
+          red_tv, pos, ComputeAtMode::BestEffort);
+    }
+  }
+}
+
+namespace {
+struct id_lt {
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/scheduler/reduction_utils.cpp:491]: Sorting function for id's before rfactor, coordinated with expectations from the inliner.

File: torch/csrc/jit/codegen/cuda/scheduler/registry.cpp [hunk 17]
```diff
@@ -638,6 +640,53 @@ std::vector<REDUCTION_OP*> findReductionOps(Fusion* fusion) {
   return red_ops;
 }
 
+std::vector<TransposeOp*> findTransposeOps(Fusion* fusion) {
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/scheduler/registry.cpp:643]: Don't allow transpose ops in reduction scheduler as it can break.

File: torch/csrc/jit/codegen/cuda/scheduler/registry.cpp [hunk 18]
```diff
@@ -638,6 +640,53 @@ std::vector<REDUCTION_OP*> findReductionOps(Fusion* fusion) {
...
+  std::vector<TransposeOp*> transpose_ops;
+  for (auto expr : fusion->exprs()) {
+    if (auto transpose_op = dynamic_cast<TransposeOp*>(expr)) {
+      transpose_ops.push_back(transpose_op);
+    }
+  }
+  return transpose_ops;
+}
+
+static bool checkPatternEquivalence(
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/scheduler/registry.cpp:653]: Moved since normalization and reduction scheduling uses it now.

File: torch/csrc/jit/codegen/cuda/scheduler/registry.cpp [hunk 19]
```diff
@@ -667,9 +716,9 @@ std::vector<REDUCTION_OP*> findReductionOps(Fusion* fusion) {
 //!        This function will be called when compiling a kernel. It should apply
 //!        scheduling to the given fusion
 
-class SingleReductionScheduler : public SchedulerEntry {
+class ReductionScheduler : public SchedulerEntry {
  public:
-  explicit SingleReductionScheduler(
+  explicit ReductionScheduler(
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/scheduler/registry.cpp:721]: No longer single reduction scheduler, general reduction scheduler if reduction patterns match.

File: torch/csrc/jit/codegen/cuda/scheduler/utils.cpp [hunk 20]
```diff
@@ -456,631 +608,6 @@ std::vector<TensorView*> getReductionTvs(Fusion* fusion) {
   return reduction_tvs;
 }
 
-TensorView* scheduleReductionTV(
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/scheduler/utils.cpp:459]: The below functions moved to reduction utils file.

File: torch/csrc/jit/codegen/cuda/scheduler/utils.cpp [hunk 21]
```diff
@@ -1576,14 +864,14 @@ std::vector<int64_t> mappedInputsOutputs(TensorView* reference_tv) {
   ca_index_map.build(fusion);
 
   auto ref_root_domain = reference_tv->getMaybeRFactorDomain();
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/scheduler/utils.cpp:871]: Changes to fix 2D pointwise scheduling logic.

File: torch/csrc/jit/codegen/cuda/transform_rfactor.cpp [hunk 22]
```diff
@@ -55,15 +55,15 @@ class ReplayRFactor : public ReplayTransformations {
     IterDomain* ido = new IterDomain(
         new Int(0),
         s->innerSplit() ? remainder->as<Int>() : s->factor(),
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/transform_rfactor.cpp:57]: Make sure rfactor preserves parallelization schemes correctly.

File: test/cpp/jit/test_gpu.cpp [hunk 23]
```diff
@@ -11082,38 +11080,23 @@ TEST(NVFuserTest, FusionIssue367_CUDA) {
...
-      {6, 4},
-  });
-
-  tv7->reorder({
-      {2, -2},
-      {3, -1},
-      {-2, 2},
-      {-1, 3},
-  });
-
```
Comment (naoyam) [inline test/cpp/jit/test_gpu.cpp:11116]: 👍

File: torch/csrc/jit/codegen/cuda/runtime/welford.cu (hunk 6 above)
Comment (naoyam) [reply torch/csrc/jit/codegen/cuda/runtime/welford.cu:21]: Do we still need to keep the type of `b_N` as a template type? Can we change the type to just `nvfuser_index_t`?

File: torch/csrc/jit/codegen/cuda/scheduler/utils.cpp [hunk 24]
```diff
@@ -16,6 +13,120 @@ namespace jit {
 namespace fuser {
 namespace cuda {
 namespace scheduler_utils {
+
+// Returns number of "valid" dimensions. e.g. if we have [I1, R2, I3, I4,
+// dontmerge[1]] result domain should be: [I1, I3*I4, R2, dontmerge[1]] with
+// return value 3
+// if we have [R1, I2, R3, I4, R4, dontmerge[0], dontmerge[1]] result domain
+// should be: [I2*I4, R1*R3, R4, dontmerge[0], dontmerge[1]] with return value 3
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/scheduler/utils.cpp]: Is `dontmerge[0]` indicating `R1` should not be merged? If so, why there is `R1*R3`?

File: torch/csrc/jit/codegen/cuda/scheduler/reduction_utils.h [hunk 25]
```diff
@@ -0,0 +1,50 @@
...
+    std::vector<std::pair<TensorView*, TensorView*>> cached_outputs);
+
+// Sort and rfactor the reference tv in a consistent way for reduction inliner.
+// Order of the sort is:
+//
+// [i-block dims, i-thread dims, i-non-constant sized, i-constant sized,
+//  r-block dims, r-thread dims, r-non-constant sized, r-constant sized,
+//  i/r-unswitched, i/r-unroll/vectorized, broadcasted dims, trivial reductions]
+//
+// Rfactored axes are reductions bound to grid or blocks. If non are bound t o
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/scheduler/reduction_utils.h]: none are bound to?

File: torch/csrc/jit/codegen/cuda/scheduler/reduction.cpp [hunk 26]
```diff
@@ -165,100 +170,103 @@ ReductionParams innerReductionHeuristic(
...
-
   // Unroll amount
-  int64_t unroll_factor = 1;
+  int64_t inner_reduction_unroll_factor = 1;
+  int64_t iter_unroll_factor = 1;
 
+  inner_reduction_unroll_factor =
+      std::min(total_reduction_numel, target_unroll);
+
+  // Grab what we can out of reduction domain, but don't go over a warp size yet
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/scheduler/reduction.cpp:180]: Duplicated comment?

File: torch/csrc/jit/codegen/cuda/scheduler/reduction.cpp [hunk 27]
```diff
@@ -165,100 +170,103 @@ ReductionParams innerReductionHeuristic(
...
-  bdimx = std::min(num_elems_in_reduction, (int64_t)warp_size);
+  bdimx = std::min(
+      std::max(
+          ceilDiv(inner_most_dimension_numel, inner_reduction_unroll_factor),
+          (int64_t)warp_size),
+      max_threads_in_block);
+  bdimx = bdimx > warp_size ? bdimx - bdimx % warp_size
+                            : scheduler_utils::lastPow2(bdimx);
+
+  // Put everything else in bdimy for now
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/scheduler/reduction.cpp:189]: ditto

File: torch/csrc/jit/codegen/cuda/scheduler/reduction.cpp [hunk 28]
```diff
@@ -668,44 +744,32 @@ TORCH_CUDA_CU_API c10::optional<ReductionParams> getReductionHeuristics(
...
-      max_dtype_size =
-          std::max(max_dtype_size, dataTypeSize(inp->getDataType().value()));
-      n_tensor_inputs++;
-    }
-  }
-
+  auto tv_inps = ir_utils::filterByType<TensorView>(fusion->inputs());
   TORCH_INTERNAL_ASSERT(
-      n_tensor_inputs > 0,
+      std::distance(tv_inps.begin(), tv_inps.end()) > 0,
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/scheduler/reduction.cpp]: nit: `!tv_inps.empty()`

File: torch/csrc/jit/codegen/cuda/compute_at.cpp [hunk 29]
```diff
@@ -400,13 +501,6 @@ unsigned int ComputeAt::backwardComputeAt_impl(
     }
 
     consumer->setMaxProducer(consumer_compute_at_pos);
-    for (auto other_consumer : ir_utils::consumerTvsOf(producer)) {
-      if (other_consumer != consumer) {
-        auto max_consumer_pos =
-            getConsumerPosAlignedToProducerCA(other_consumer, producer);
-        other_consumer->setMaxProducer(max_consumer_pos);
-      }
-    }
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/compute_at.cpp:409]: Why is this deleted?

File: torch/csrc/jit/codegen/cuda/compute_at.cpp [hunk 30]
```diff
@@ -348,6 +348,96 @@ void ComputeAt::runWith(
...
+    const TensorView* consumer,
+    int consumer_or_producer_pos,
+    bool consumer_pos = true) {
+  FUSER_PERF_SCOPE("transform_replay.cpp::skipReplay");
+
+  const auto c2p_root_map =
+      PairwiseRootDomainMap(producer, consumer)
+          .mapConsumerToProducer(consumer->domain(), producer->domain());
+
+  // IterDomains in consumer root not in producer root
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/compute_at.cpp]: nit: not `not in producer root`

**Comment by naoyam** [comment]
Looked at almost all files, except for fusion segmentor, which I leave to Shiming.

File: torch/csrc/jit/codegen/cuda/fusion_segmenter.cpp [hunk 31]
```diff
@@ -758,21 +758,31 @@ void SegmentedFusion::finalize() {
...
-    std::unordered_set<Val*> input_group_set(
-        input_group_vec.begin(), input_group_vec.end());
 
-    auto expr_set = DependencyCheck::getAllExprsBetween(
-        input_group_set, getAllOutputs(group));
-    group->exprs_ = std::vector<Expr*>(expr_set.begin(), expr_set.end());
+      // Need a valid expression list to continue. Update from and to group.
+      // TODO : this could have been a general operation that
+      //  the group supports. Could consider moving this into
+      //  segmentedGroup in a follow up.
```
Comment (shmsong) [inline torch/csrc/jit/codegen/cuda/fusion_segmenter.cpp:765]: Thanks for fixing this 👍  I will put this functionality inside `SegmentedGroup` in a follow up. 

I assume the segfault comes from iterating through the expressions in groups that already had updated boundary, but will double check.

File: torch/csrc/jit/codegen/cuda/scheduler/compile_time_info.h [hunk 32]
```diff
@@ -89,15 +83,15 @@ class ScopePersistentFactorInfo {
       CompileTimeEntryType::SCOPE_PERSISTENT_FACTOR_INFO;
 };
 
-//! Entry type definition class for `MAPPED_INPUTS_OUTPUTS`,
+//! Entry type definition class for `BROADCAST_BYTE_MULTIPLES`,
 //!  stores number of inputs/outputs non-broadcast iterdomain
```
Comment (shmsong) [inline torch/csrc/jit/codegen/cuda/scheduler/compile_time_info.h]: The description doesn't seem to match `BROADCAST_BYTE_MULTIPLES`

File: torch/csrc/jit/codegen/cuda/compute_at.cpp (hunk 29 above)
Comment (csarofeen) [reply torch/csrc/jit/codegen/cuda/compute_at.cpp:409]: I can't remember clearly, but I believe the issue is that we're going to replay other consumers anyways because if there's not a common consumer, we will propagate through those paths anyways. I'm confident it's unnecessary, but I think it as causing some unexpected behavior.

File: torch/csrc/jit/codegen/cuda/runtime/welford.cu (hunk 6 above)
Comment (csarofeen) [reply torch/csrc/jit/codegen/cuda/runtime/welford.cu:21]: What do you think @shmsong?

File: torch/csrc/jit/codegen/cuda/scheduler/reduction.cpp (hunk 28 above)
Comment (csarofeen) [reply torch/csrc/jit/codegen/cuda/scheduler/reduction.cpp]: lol, what was I doing? thanks.

File: torch/csrc/jit/codegen/cuda/scheduler/utils.cpp (hunk 24 above)
Comment (csarofeen) [reply torch/csrc/jit/codegen/cuda/scheduler/utils.cpp]: Updated comment.

**Comment by csarofeen** [comment]
@shmsong thoughts on changing welford runtime file?

File: torch/csrc/jit/codegen/cuda/compute_at.cpp (hunk 29 above)
Comment (csarofeen) [reply torch/csrc/jit/codegen/cuda/compute_at.cpp:409]: I put it back to test, seems it's just unnecessary, it's not causing issues.

Comment (naoyam) [reply torch/csrc/jit/codegen/cuda/compute_at.cpp:409]: Hmm, it seems unnecessary, but wasn't there some cases it was actually necessary? Here's the PR that introduced this logic. #1035

File: torch/csrc/jit/codegen/cuda/runtime/welford.cu (hunk 6 above)
Comment (shmsong) [reply torch/csrc/jit/codegen/cuda/runtime/welford.cu:21]: I think it is a good idea to unify `TN` here and `nvfuser_index_t`, if we will not try to handle running stats between kernels with welford op. 

A few other things will need to update to accommodate this change though:

1. `nvfuser_index_t` setting need to be more conservative by another 1 bit to make sure both `size` and `index` fit in, which we probably don't mind in the general case. 

2. We probably also want to update the workspace and welford op creation a bit so we don't always allocate `int64_t` in register and shared mem. 

I can do these in a follow up and remove `TN` if that sounds ok.

**Review by shmsong — COMMENTED** [review]
Still going through the scheduler and registry changes but had some comments/responses on the welford and segmenter.

File: torch/csrc/jit/codegen/cuda/scheduler/pointwise.cpp [hunk 33]
```diff
@@ -67,32 +67,20 @@ c10::optional<PointwiseParams> getPointwiseHeuristics(
...
-    // Create empty entries for vectorizable inputs outputs
-    //  and mapping count
     auto vectorizable_inputs_outputs_entry = HeuristicSummaryEntry<
```
Comment (shmsong) [inline torch/csrc/jit/codegen/cuda/scheduler/pointwise.cpp:70]: Probably missed this last time. Should we return early under this case? 

These empty entries will be overwritten by line 145 if there's no early return. Or maybe we could change `HeuristicSummaryCache` to `write once` behavior if needed in this case.

File: torch/csrc/jit/codegen/cuda/scheduler/pointwise.cpp [hunk 34]
```diff
@@ -67,32 +67,20 @@ c10::optional<PointwiseParams> getPointwiseHeuristics(
...
-    TORCH_INTERNAL_ASSERT(
-        inferred_val.has_value(),
-        "Error inferring size for pointwise scheduler.");
-    elem_counts[ref_i] = inferred_val.value();
-    n_elems *= inferred_val.value();
+        HeuristicCompileTime::VectorizableInputsAndOutputs>(
+        data_cache, [&largest_out]() {
+          return std::make_unique<std::vector<TensorView*>>();
+        });
+    vectorizable_inputs_outputs_entry.get();
```
Comment (shmsong) [inline torch/csrc/jit/codegen/cuda/scheduler/pointwise.cpp:74]: `.get()` would just return a reference without any side-effect. Please let me know if there's any stateful change was expected here and I could add API call for that on `HeuristicSummaryCache`

File: torch/csrc/jit/codegen/cuda/scheduler/reduction.cpp [hunk 35]
```diff
@@ -142,7 +142,12 @@ ReductionParams innerReductionHeuristic(
     // targetting 4 waves, so try to use a quarter of available threads
     max_threads_in_block = std::min(
         ceilDiv(n_elems, target_blocks * target_unroll),
-        ceilDiv(device_max_threads_per_multiprocessor, (int64_t)4));
+        ceilDiv(device_max_threads_per_multiprocessor, (int64_t)8));
```
Comment (shmsong) [inline torch/csrc/jit/codegen/cuda/scheduler/reduction.cpp:145]: nitpick/question: comment says `targeting 4 waves` does this change mean targeting 8 waves instead?

File: torch/csrc/jit/codegen/cuda/scheduler/registry.cpp [hunk 36]
```diff
@@ -1114,17 +1192,21 @@ void HeuristicSummary::validate() const {
...
+          entry_type_map_.count(EntryType::BROADCAST_BYTE_MULTIPLES));
       break;
     case ScheduleHeuristic::Reduction:
       TORCH_INTERNAL_ASSERT(entry_type_map_.count(EntryType::REDUCTION_TVS));
       TORCH_INTERNAL_ASSERT(
           entry_type_map_.count(EntryType::VECTORIZABLE_INPUTS_AND_OUTPUTS));
+      TORCH_INTERNAL_ASSERT(
+          entry_type_map_.count(EntryType::UNROLLABLE_INPUTS_AND_OUTPUTS));
       break;
     case ScheduleHeuristic::Normalization:
```
Comment (shmsong) [inline torch/csrc/jit/codegen/cuda/scheduler/registry.cpp]: nitpick: do we want to rename `ScheduleHeuristic::Normalization` to `ScheduleHeuristic::Persistent` ?

File: torch/csrc/jit/codegen/cuda/fusion_segmenter.cpp (hunk 31 above)
Comment (csarofeen) [reply torch/csrc/jit/codegen/cuda/fusion_segmenter.cpp:765]: Yes, that's the case. When modifying the expressions they become invalid, and I was later trying to access them. No idea how this didn't show up before.

File: torch/csrc/jit/codegen/cuda/scheduler/pointwise.cpp (hunk 33 above)
Comment (csarofeen) [reply torch/csrc/jit/codegen/cuda/scheduler/pointwise.cpp:70]: I think we should be able to just return early, checking if there's any failing tests from this.

File: torch/csrc/jit/codegen/cuda/scheduler/reduction.cpp (hunk 35 above)
Comment (csarofeen) [reply torch/csrc/jit/codegen/cuda/scheduler/reduction.cpp:145]: Thanks, I didn't mean to leave that change in.

File: torch/csrc/jit/codegen/cuda/fusion_segmenter.cpp (hunk 31 above)
Comment (shmsong) [reply torch/csrc/jit/codegen/cuda/fusion_segmenter.cpp:765]: Thanks. I will take a deeper look when doing the follow up.

File: torch/csrc/jit/codegen/cuda/compute_at.cpp (hunk 29 above)
Comment (csarofeen) [reply torch/csrc/jit/codegen/cuda/compute_at.cpp:409]: In that PR: https://github.com/csarofeen/pytorch/commit/ad070a2df94f7ee2429dcfd66d05404bb8cdff88#diff-a09f5a841235343227ca4aa5021932ff8ef7c883578679bb366652ba0c18884cR616-R647

the important part there is:
```.cpp
    if (new_consummer_pa_pos < running_consumer->getMaxProducerPosition()) {
      running_consumer->setMaxProducer(new_consummer_pa_pos, true);
    }
```

Now we have:
```.cpp
void ComputeAt::hoistInnermostBroadcast() {
  auto fusion = producer_->fusion();

  std::unordered_set<TensorView*> consumers_to_update;

  auto all_vals = fusion->usedMathVals();
  auto all_tvs = ir_utils::filterByType<TensorView>(all_vals);

  for (auto running_producer : all_tvs) {
    if (!running_producer->isFusionInput()) {
      auto producer_ca_pos = running_producer->getComputeAtPosition();
      // Find the innermost iterdomain that is not a broadcast
      auto new_ca_pos = getInnermostNonBroadcastIdFrom(running_producer);
      // Update the compute at pos of this producer if the original
      //  compute at is within inner most broadcast axes
      if (new_ca_pos < producer_ca_pos) {
        running_producer->setComputeAt(new_ca_pos, true);
      }
      // Mark all consumers of this producer for later produce
      //  position update.
      // This is safe with segmented fusion. TV uses will reset
      //  when FusionSegmentGuard try to change the IO.
      auto tv_consumers = ir_utils::consumerTvsOf(running_producer);
      consumers_to_update.insert(tv_consumers.begin(), tv_consumers.end());
    }
  }

  // Update the produce positions of all affected consumers
  for (auto running_consumer : consumers_to_update) {
    TORCH_INTERNAL_ASSERT(running_consumer->definition() != nullptr);
    resetMaxProducerPos(running_consumer);
  }
}

Along with...
```.cpp
void ComputeAt::resetMaxProducerPos(TensorView* consumer_tv) {
  if (consumer_tv->definition() == nullptr) {
    consumer_tv->setMaxProducer(0, true);
  }

  unsigned int new_consummer_pa_pos = 0;

  // Re-compute the max producer position as one or more
  //  of the producers of this consumer have updated their
  //  compute at position.
  for (auto inp : ir_utils::producerTvsOf(consumer_tv)) {
    if (!inp->isFusionInput()) {
      // Locate consumer's position that aligns with
      //  the producer's new compute at axis.
      unsigned int inp_ca_pos_to_consumer =
          getConsumerPosAlignedToProducerCA(consumer_tv, inp);

      // Populate the max consumer position required by
      //  producer compute at.
      new_consummer_pa_pos =
          std::max(new_consummer_pa_pos, inp_ca_pos_to_consumer);
    }
  }

  consumer_tv->setMaxProducer(new_consummer_pa_pos, true);
}
```

updating `consumer_tv` max producer position is no longer conditional, it really is happening all the time now as part of cleanup. We're doing what we were previously, just not during the compute at pass anymore.

**Review by naoyam — APPROVED** [review]
